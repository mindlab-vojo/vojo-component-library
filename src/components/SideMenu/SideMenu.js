import React, { useState } from 'react'
import PropTypes from 'prop-types'

import IconManager from '../IconManager'
import Typography from '../Typography'

import ThemeConsumer from '../../style/ThemeManager/ThemeProvider'
import * as SideMenu from './SideMenu.style'
import Button from '../Button'

const SideMenuComponent = ({
  buttons,
  items,
  onClickClose,
  onClickOutside,
  terms
}) => {

  const renderItems = (theme) => {
    return items ? 
      items.map((item, index) => (
        <SideMenu.Item
          onClick={() => item.func && item.func()}
          key={index}>
          <SideMenu.ItemIcon>
            <IconManager
              fill={theme.colors.primaryColor}
              icon={item.icon}
              width="22px"
            />
          </SideMenu.ItemIcon>
          <SideMenu.ItemText>
            <Typography
              color={theme.colors.primaryColor}
              fontSize="18px"
              fontWeight="bolder">
              {item.text}
            </Typography>
          </SideMenu.ItemText>
        </SideMenu.Item>
      )) : null
  }

  const renderButtons = (theme) => {
    return buttons ? 
      buttons.map((button, index) => (
        <SideMenu.Button
          key={index}>
          <SideMenu.ButtonWrapper>
            <Button
              onButtonClick={() => button.func && button.func()}
              backgroundColor={theme.colors.whiteColor}
              borderColor={theme.colors.primaryColor}
              hoverColor={theme.colors.whiteColor}
              id={`${index}-menu-button`}
              name={`${index}-menu-button`}>
              <Typography
                color={theme.colors.primaryColor}
                fontSize="18px"
                fontWeight="bolder">
                {button.text}
              </Typography>
            </Button>
          </SideMenu.ButtonWrapper>
        </SideMenu.Button>
      )) : null
  }

  const renderComponent = (theme) => (
    <SideMenu.Wrapper
      className="SideMenu">
      <SideMenu.Background
        className="SideMenu__Background"
        onClick={onClickOutside}/>
      <SideMenu.Menu
        className="SideMenu__Menu">
        <SideMenu.Close
          className="SideMenu__CloseWrapper">
          <SideMenu.CloseIcon
            className="SideMenu__Close"
            onClick={onClickClose}>
            <IconManager
              fill={theme.colors.blackColor}
              icon="Close"
              width="25px"
            />
          </SideMenu.CloseIcon>
        </SideMenu.Close>
        <SideMenu.MenuItems>
          {renderItems(theme)}
        </SideMenu.MenuItems>
        <SideMenu.MenuButtons>
          {renderButtons(theme)}
        </SideMenu.MenuButtons>
        {
          terms && (
            <SideMenu.Terms
              onClick={terms.func}>
              <Typography
                isUnderlined>
                {terms.text}
              </Typography>
            </SideMenu.Terms>
          )
        }
      </SideMenu.Menu>
    </SideMenu.Wrapper>
  )

  return (
    <ThemeConsumer>
      {({ theme }) => renderComponent(theme)}
    </ThemeConsumer>
  )
}

SideMenuComponent.propTypes = {
  buttons: PropTypes.array,
  items: PropTypes.array,
  onClickClose: PropTypes.func,
  onClickOutside: PropTypes.func,
  terms: PropTypes.object
}

export default SideMenuComponent
