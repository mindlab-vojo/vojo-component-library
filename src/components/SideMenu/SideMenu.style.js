import styled from 'styled-components'

export const Wrapper = styled.div`
  height: 100vh;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 4000;
`

export const Background = styled.div`
  background-color: #000000;
  height: 100vh;
  left: 0;
  opacity:0.2;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 4001;
`

export const Menu = styled.div`
  background-color: #FFFFFF;
  box-sizing: border-box;
  height: 100vh;
  left: 0;
  max-width: 360px;
  padding: 25px;
  position: absolute;
  overflow-y: scroll;
  top: 0;
  width: 100%;
  z-index: 4002;

  &::-webkit-scrollbar {
    display: none;
  };
`

export const Close = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-bottom: 30px;
  width: 100%;
`

export const CloseIcon = styled.div`
  cursor: pointer;
  width: fit-content;
`

export const MenuItems = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`

export const Item = styled.div`
  align-items: center;
  cursor: pointer;
  display: flex;
  margin: 7px 0px;
  width: 100%;

  &:first-of-type {
    margin-top: 0;
  };

  &:last-of-type {
    margin-bottom: 0;
  };
`

export const ItemIcon = styled.div`
  display: flex;
  width: 22px;
`

export const ItemText = styled.div`
  display: flex;
  margin-left: 15px;
  width: 100%;
`

export const MenuButtons = styled.div`
  display: flex;
  flex-direction: column;
  margin-top: 50px;
  width: 100%;
`

export const Button = styled.div`
  align-items: center;
  cursor: pointer;
  display: flex;
  margin: 7px 0px;
  width: 100%;

  &:first-of-type {
    margin-top: 0;
  };

  &:last-of-type {
    margin-bottom: 0;
  };
`

export const ButtonWrapper = styled.div`
  width: 128px;
`

export const Terms = styled.div`
  align-items: center;
  cursor: pointer;
  box-sizing: border-box;
  bottom: 20px;
  display: flex;
  justify-content: center;
  flex-direction: column;
  position: absolute;
`