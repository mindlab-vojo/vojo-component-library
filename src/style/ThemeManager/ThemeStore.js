import { vojo } from './../themes'

class ThemeStore {

  constructor(theme) {
    this.defaultTheme = vojo
    this.theme = theme ? theme : this.defaultTheme
  }

  configure = ({ theme }) =>  {
    if (theme) this.configureTheme(theme)
    else this.configureTheme(this.defaultTheme)
  }

  configureTheme = (theme) => {
    this.theme = theme
  }

  modify = (modifiedProps) => {
    const currentTheme = this.theme
    return { currentTheme, modifiedProps }
  }

  getTheme = () => this.theme

}

export default ThemeStore