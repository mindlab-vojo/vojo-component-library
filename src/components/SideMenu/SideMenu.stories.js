import React from 'react';
import { withKnobs, object, number, boolean } from '@storybook/addon-knobs';
import { withA11y } from '@storybook/addon-a11y';
import { action } from '@storybook/addon-actions'

import SideMenu from '.'; 
import { ThemeProvider } from "../../../src";
import categories from '../../../.storybook/categories';

export const menuButtons = [
  {
    text: 'Sair',
    func: () => console.log('click')
  },
]

export const menuItems = [
  {
    icon: 'Check',
    text: 'Oportunidades',
    func: () => console.log('click')
  },
  {
    icon: 'Check',
    text: 'Cursos Gratuitos',
    func: () => console.log('click')
  },
  {
    icon: 'Check',
    text: 'Candidaturas Realizadas',
    func: () => console.log('click')
  },
  {
    icon: 'Check',
    text: 'Página de Ajuda',
    func: () => console.log('click')
  },
  {
    icon: 'Check',
    text: 'Minha Conta',
    func: () => console.log('click')
  },
]

export const termsObject = {
  text: 'termos de uso e política de privacidade',
  func: () => console.log('click')
}

export const sideMenu = () => {  
  const animationDuration = number('animationDuration', 0.3)
  const buttons = object('buttons', menuButtons)
  const items = object('items', menuItems)
  const onClickClose = action('onClickClose')
  const onClickOutside = action('onClickOutside')
  const show = boolean('show', false)
  const terms = object('terms', termsObject)

  return (
    <SideMenu
      animationDuration={animationDuration}
      buttons={buttons}
      items={items}
      onClickClose={onClickClose}
      onClickOutside={onClickOutside}
      show={show}
      terms={terms}
    />
  )
}

export default {
  title: categories.Headers,
  name: 'SideMenu',
  component: sideMenu,
  decorators: [
    withKnobs, 
    withA11y, 
    (story) => <ThemeProvider>{ story() }</ThemeProvider>,
    (story) => <div style={{ display: "flex" }}>{ story() }</div>,
  ],
}
  