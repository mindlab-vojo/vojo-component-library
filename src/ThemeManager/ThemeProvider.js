import React, { Component, createContext } from 'react'
import PropTypes from 'prop-types'

import ThemeStore from './ThemeStore';

// Provider and Consumer instances
const { Provider, Consumer } = createContext();

class ThemeProvider extends Component {
  static propTypes = {
    children: PropTypes.node,
    theme: PropTypes.string
  }

  state = {
    themeConfig: new ThemeStore()
  }

  componentDidMount() {
    if (this.props.theme) this.setTheme()
  }

  componentDidUpdate(prevProps) {
    if (this.props.theme !== prevProps.theme) this.setTheme()
  }

  setTheme = () => {
    const { theme } = this.props
    const { themeConfig } = this.state
    const updatedTheme = themeConfig.configure({ theme })
    this.setState({ theme: updatedTheme })
  }

  render() {
    return (
      <Provider value={{ theme: this.state.themeConfig.theme }}>
        {this.props.children}
      </Provider>
    )
  }
}

export { ThemeProvider }

export default Consumer