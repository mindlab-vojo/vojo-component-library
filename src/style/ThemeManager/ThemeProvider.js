import React, { Component, createContext } from 'react'
import PropTypes from 'prop-types'

import ThemeStore from './ThemeStore';

// Provider and Consumer instances
const { Provider, Consumer } = createContext();

class ThemeProvider extends Component {
  static propTypes = {
    children: PropTypes.node,
    theme: PropTypes.string
  }

  state = {
    themeConfig: new ThemeStore()
  }

  componentDidMount() {
    if (this.props.theme) this.setTheme()
  }

  componentDidUpdate(prevProps) {
    if (this.props.theme !== prevProps.theme) this.setTheme()
  }

  setTheme = () => {
    const { theme } = this.props
    const { themeConfig } = this.state
    const updatedTheme = themeConfig.configure({ theme })
    this.setState({ theme: updatedTheme })
  }

  getThemeFont = () => {
    const { themeConfig } = this.state
    return themeConfig.font
  }

  render() {

    const GlobalFont = this.getThemeFont()

    return (
      <Provider value={{ theme: this.state.themeConfig.theme }}>
        <GlobalFont/>
        {this.props.children}
      </Provider>
    )
  }
}

export { ThemeProvider }

export default Consumer