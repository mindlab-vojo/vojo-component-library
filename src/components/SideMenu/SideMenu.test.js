import React from 'react';
import { render } from '@testing-library/react';

import SideMenu from './';
import { ThemeProvider } from './../../style/ThemeManager/ThemeProvider';

test('Test SideMenu component render', () => {
  const fakeFn = jest.fn()
  const component = render(
    <ThemeProvider>
      <SideMenu 
        animationDuration={1}
        buttons={[
          {
            text: 'Sair',
            func: () => console.log('click')
          },
        ]}
        items={[
          {
            icon: 'Suitcase',
            text: 'Oportunidades',
            func: () => console.log('click')
          },
          {
            icon: 'Course',
            text: 'Cursos Gratuitos',
            func: () => console.log('click')
          }
        ]}
        onClickClose={fakeFn}
        onClickOutside={fakeFn}
        show={true}
        terms={{
          text: 'termos de uso e polÃ­tica de privacidade',
          func: () => console.log('click')
        }}/>
    </ThemeProvider>
  );
  expect(component).toMatchSnapshot()
});
