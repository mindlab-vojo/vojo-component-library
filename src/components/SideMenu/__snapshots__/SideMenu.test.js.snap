// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SideMenu component render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ dulYen SideMenu"
      >
        <div
          class="sc-gsTCUz fVGUcF SideMenu__Background"
        />
        <div
          class="sc-dlfnbm kzGwTK SideMenu__Menu"
        >
          <div
            class="sc-hKgILt fBlSnY SideMenu__CloseWrapper"
          >
            <div
              class="sc-eCssSg fvFtc SideMenu__Close"
            >
              <svg
                height="24"
                viewBox="0 0 24 24"
                width="25px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  fill="#212529"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </div>
          </div>
          <div
            class="sc-jSgupP ceWOEk"
          >
            <div
              class="sc-gKsewC dwIoxt"
            >
              <div
                class="sc-iBPRYJ gdWKsf"
              >
                <svg
                  height="19"
                  viewBox="0 0 20 19"
                  width="22px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 4H14V2C14 0.89 13.11 0 12 0H8C6.89 0 6 0.89 6 2V4H2C0.89 4 0.00999999 4.89 0.00999999 6L0 17C0 18.11 0.89 19 2 19H18C19.11 19 20 18.11 20 17V6C20 4.89 19.11 4 18 4ZM12 4H8V2H12V4Z"
                    fill="#391DDD"
                  />
                </svg>
              </div>
              <div
                class="sc-fubCfw hzOVYV"
              >
                <p
                  class="sc-dQppl kVMoQj"
                >
                  Oportunidades
                </p>
              </div>
            </div>
            <div
              class="sc-gKsewC dwIoxt"
            >
              <div
                class="sc-iBPRYJ gdWKsf"
              >
                <svg
                  height="18"
                  viewBox="0 0 22 18"
                  width="22px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4 10.18V14.18L11 18L18 14.18V10.18L11 14L4 10.18ZM11 0L0 6L11 12L20 7.09V14H22V6L11 0Z"
                    fill="#391DDD"
                  />
                </svg>
              </div>
              <div
                class="sc-fubCfw hzOVYV"
              >
                <p
                  class="sc-bqyKva gzNxIg"
                >
                  Cursos Gratuitos
                </p>
              </div>
            </div>
          </div>
          <div
            class="sc-pFZIQ GIcVx"
          >
            <div
              class="sc-jrAGrp FHQqT"
            >
              <div
                class="sc-kEjbxe gShzib"
              >
                <button
                  class="sc-crrsfI jgdrYy Button"
                  id="0-menu-button"
                  name="0-menu-button"
                  type="button"
                  width="100%"
                >
                  <p
                    class="sc-kstrdz fzlLig"
                  >
                    Sair
                  </p>
                </button>
              </div>
            </div>
          </div>
          <div
            class="sc-iqHYGH dSQeZq"
          >
            <p
              class="sc-hBEYos fJcqRM"
            >
              termos de uso e política de privacidade
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ dulYen SideMenu"
    >
      <div
        class="sc-gsTCUz fVGUcF SideMenu__Background"
      />
      <div
        class="sc-dlfnbm kzGwTK SideMenu__Menu"
      >
        <div
          class="sc-hKgILt fBlSnY SideMenu__CloseWrapper"
        >
          <div
            class="sc-eCssSg fvFtc SideMenu__Close"
          >
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="25px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                fill="#212529"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
        </div>
        <div
          class="sc-jSgupP ceWOEk"
        >
          <div
            class="sc-gKsewC dwIoxt"
          >
            <div
              class="sc-iBPRYJ gdWKsf"
            >
              <svg
                height="19"
                viewBox="0 0 20 19"
                width="22px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 4H14V2C14 0.89 13.11 0 12 0H8C6.89 0 6 0.89 6 2V4H2C0.89 4 0.00999999 4.89 0.00999999 6L0 17C0 18.11 0.89 19 2 19H18C19.11 19 20 18.11 20 17V6C20 4.89 19.11 4 18 4ZM12 4H8V2H12V4Z"
                  fill="#391DDD"
                />
              </svg>
            </div>
            <div
              class="sc-fubCfw hzOVYV"
            >
              <p
                class="sc-dQppl kVMoQj"
              >
                Oportunidades
              </p>
            </div>
          </div>
          <div
            class="sc-gKsewC dwIoxt"
          >
            <div
              class="sc-iBPRYJ gdWKsf"
            >
              <svg
                height="18"
                viewBox="0 0 22 18"
                width="22px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 10.18V14.18L11 18L18 14.18V10.18L11 14L4 10.18ZM11 0L0 6L11 12L20 7.09V14H22V6L11 0Z"
                  fill="#391DDD"
                />
              </svg>
            </div>
            <div
              class="sc-fubCfw hzOVYV"
            >
              <p
                class="sc-bqyKva gzNxIg"
              >
                Cursos Gratuitos
              </p>
            </div>
          </div>
        </div>
        <div
          class="sc-pFZIQ GIcVx"
        >
          <div
            class="sc-jrAGrp FHQqT"
          >
            <div
              class="sc-kEjbxe gShzib"
            >
              <button
                class="sc-crrsfI jgdrYy Button"
                id="0-menu-button"
                name="0-menu-button"
                type="button"
                width="100%"
              >
                <p
                  class="sc-kstrdz fzlLig"
                >
                  Sair
                </p>
              </button>
            </div>
          </div>
        </div>
        <div
          class="sc-iqHYGH dSQeZq"
        >
          <p
            class="sc-hBEYos fJcqRM"
          >
            termos de uso e política de privacidade
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
